processors/src/VertexAnaProcessor.cxx

```diff
@@ -322,10 +350,81 @@ bool VertexAnaProcessor::process(IEvent* ievent) {
+                }
+                //std::cout << "There are " << ele_trk_hits->GetEntries() << " hits on this track" << std::endl;
+                //Count the number of hits per part on the track
+                std::map<int, int> nHits4part;
+                for(int i = 0; i < ele_trk_hits->GetEntries(); i++)
+                {
+                    TrackerHit* eleHit = (TrackerHit*)ele_trk_hits->At(i);
+                    for(int idI = 0; idI < trueHitIDs[eleHit->getID()].size(); idI++ )
+                    {
+                        int partID = trueHitIDs[eleHit->getID()].at(idI);
+                        if ( nHits4part.find(partID) == nHits4part.end() ) 
+                        {
+                            // not found
+                            nHits4part[partID] = 1;
+                        } 
+                        else 
+                        {
+                            // found
+                            nHits4part[partID]++;
+                        }
+                    }
+                }
+
+                //Determine the MC part with the most hits on the track
+                int maxNHits = 0;
+                int maxID = 0;
+                for (std::map<int,int>::iterator it=nHits4part.begin(); it!=nHits4part.end(); ++it)
+                {
+                    if(it->second > maxNHits)
+                    {
+                        maxNHits = it->second;
+                        maxID = it->first;
+                    }
+                }
+
+                //Find the correct mc part and grab mother id
+                int isRadEle = 0;
+                int isRecEle = 0;
+                TVector3 trueEleP;
+                for(int i = 0; i < mcParts_->size(); i++)
+                {
+                    int momPDG = mcParts_->at(i)->getMomPDG();
+                    if(mcParts_->at(i)->getPDG() == 11 && momPDG == 622) 
+                    {
+                        std::vector<double> lP = mcParts_->at(i)->getMomentum();
+                        trueEleP.SetXYZ(lP[0],lP[1],lP[2]);
+                    }
+                    if(mcParts_->at(i)->getID() != maxID) continue;
+                    if(momPDG == 622) isRadEle = 1;
+                    if(momPDG == 623) isRecEle = 1;
+                }
+                double momRatio = recEleP.Mag() / trueEleP.Mag();
+                double momAngle = trueEleP.Angle(recEleP) * TMath::RadToDeg();
+                if (!_reg_vtx_selectors[region]->passCutLt("momRatio_lt", momRatio, weight)) continue;
+                if (!_reg_vtx_selectors[region]->passCutGt("momRatio_gt", momRatio, weight)) continue;
+                if (!_reg_vtx_selectors[region]->passCutLt("momAngle_lt", momAngle, weight)) continue;
+
+                if (!_reg_vtx_selectors[region]->passCutEq("isRadEle_eq", isRadEle, weight)) continue;
+                if (!_reg_vtx_selectors[region]->passCutEq("isRecEle_eq", isRecEle, weight)) continue;
+            }
+
```

Comment:
Can you give me an example of another place we would put it? Not sure it makes sense to do this anywhere beyond the driver that is analyzing vertices. Not saying an example doesn't exist, but curious what you had in mind.

Reply:
It's useful in general to have a separated tool that does the track/particle <-> MCParticle matching.

Reply:
yeah, so the track/particle matching part of this could be moved to a separate tool, agreed. I guess put it in the LCIO->ROOT utilities? This part ultimately really is doing vertex <--> MCParticle matching, so it won't be the whole thing.